src/object/cli_obj.c

```diff
@@ -617,16 +516,163 @@ obj_shard_find_replica(struct dc_object *obj, unsigned int target,
 }
 
 static int
-obj_grp_leader_get(struct dc_object *obj, int idx, unsigned int map_ver, uint8_t *bit_map)
+obj_ec_leader_select(struct dc_object *obj, int grp_idx, uint64_t dkey_hash,
+		     uint8_t *bit_map)
+{
+	struct daos_oclass_attr *oca;
+	struct pl_obj_shard	*pl_shard;
+	int			tgt_idx;
+	int			grp_size;
+	int			grp_start;
+	int			i;
+	int			shard;
+
+	oca = daos_oclass_attr_find(obj->cob_md.omd_id, NULL);
+	grp_size = obj_ec_tgt_nr(oca);
+	grp_start = grp_idx * grp_size;
+
+	/* 1. Find one from parity, and start from the last parity. */
+	tgt_idx = obj_ec_shard_idx(dkey_hash, oca, grp_size - 1);
+	for (i = 0; i < obj_ec_parity_tgt_nr(oca);
+	     i++, tgt_idx = (tgt_idx - 1 + grp_size) % grp_size) {
+
+		shard = grp_start + tgt_idx;
+		pl_shard = obj_get_shard(obj, shard);
+		if (pl_shard->po_target == -1 || pl_shard->po_shard == -1 ||
+		    pl_shard->po_rebuilding) {
+			/* Then try former one */
+			continue;
+		}
+		goto found;
+	}
+
+	/* If we do not know which data shards participate in the transaction,
+	 * let's return DER_STALE, then object I/O might refresh the pool map
+	 * and layout until parity rebuilt finish.
+	 */
+	if (bit_map == NIL_BITMAP) {
+		D_WARN(DF_OID" all parity shards %d are in rebuilding, retry later.\n",
+		       DP_OID(obj->cob_md.omd_id), obj_ec_parity_tgt_nr(oca));
+		return -DER_STALE;
+	}
+
+	/* Find one from the bitmap */
+	shard = -1;
+	tgt_idx = obj_ec_shard_idx(dkey_hash, oca, 0);
+	/* Choose one from the data shards within bit_map */
+	for (i = 0; i < obj_ec_data_tgt_nr(oca); i++, tgt_idx = (tgt_idx + 1) % grp_size) {
+		int shard_idx;
+
+		if (bit_map != NIL_BITMAP && isclr(bit_map, tgt_idx))
```

Comment:
bit_map NULL already checked at above L553

Reply:
oh, fixed  this in a different patch.